### ISSUE TITLE
cyclegan in 'wgangp' mode

### ISSUE THREAD

**Issue by pengyinxw** [junyanz/pytorch-CycleGAN-and-pix2pix#637]
Hi, I can get good results with cyclegan in 'lsgan' mode on horse2zebra dataset, but when I change the gan mode to 'wgangp' and follow your codes:
self.loss_gp, gradients = cal_gradient_penalty()...
self.loss_gp.backward(retain_graph=True)
self.loss_D = (self.loss_D_fake + self.loss_D_real) * 0.5
self.loss_D.backward()
to add loss_gp. However, the generated images are almost the same as the real images. That is the style transfer is failed. Could you pls give me some advices? Thank you in  advance.

**Comment by junyanz** [comment]
CycleGAN is implemented using LSGAN loss. I recommend that you use LSGAN loss. WGANGP loss is included for completeness, in case people use the codebase for other purposes.

**Comment by pengyinxw** [comment]
Thank you! However, we can theoretically implement CycleGAN using WGANGP loss, right?  
Btw. have you implemented CycleGAN with WGANGP loss?
If I want to improve CycleGAN, could you give me some advices what I can do ?

**Comment by junyanz** [comment]
I haven't tried the CycleGAN with WGAN-GP loss but didn't get significantly better performance. 
You may want to check out some recent work such as [1](https://arxiv.org/pdf/1808.04325.pdf).

**Comment by moh3th1** [comment]
Hi @junyanz 
I want to make sure that I am trying CycleGAN with WGAN-GP loss in the same setup and with the same configuration you tried it.

This is my modification inspired by #439 and [VON](https://github.com/junyanz/VON/blob/master/models/shape_gan_model.py#L80) on **cycle_gan_model.py**:
```
def backward_D_basic(self, netD, real, fake):
        """Calculate GAN loss for the discriminator

        Parameters:
            netD (network)      -- the discriminator D
            real (tensor array) -- real images
            fake (tensor array) -- images generated by a generator

        Return the discriminator loss.
        We also call loss_D.backward() to calculate the gradients.
        """
        # Real
        pred_real = netD(real)
        loss_D_real = self.criterionGAN(pred_real, True)
        # Fake
        pred_fake = netD(fake.detach())
        loss_D_fake = self.criterionGAN(pred_fake, False)
        # wgan-gp
        gradient_penalty, gradients = networks.cal_gradient_penalty(
            netD, real, fake, self.device,lambda_gp=10.0
        )
        gradient_penalty.backward(retain_graph=True)
        # Combined loss and calculate gradients
        loss_D = (loss_D_real + loss_D_fake) * 0.5
        loss_D.backward()
        return loss_D
```
and to start training  
```
python train.py --dataroot ./datasets/maps --name maps_cyclegan --model cycle_gan --gan_mode wgangp --norm instance 
```
Is this all?  or should I take care of other things?
Is this the same setup you tried ?

**Comment by junyanz** [comment]
I think so.

**Comment by Haicang** [comment]
> Hi @junyanz
> I want to make sure that I am trying CycleGAN with WGAN-GP loss in the same setup and with the same configuration you tried it.
> 
> This is my modification inspired by #439 and [VON](https://github.com/junyanz/VON/blob/master/models/shape_gan_model.py#L80) on **cycle_gan_model.py**:
> 
> ```
> def backward_D_basic(self, netD, real, fake):
>         """Calculate GAN loss for the discriminator
> 
>         Parameters:
>             netD (network)      -- the discriminator D
>             real (tensor array) -- real images
>             fake (tensor array) -- images generated by a generator
> 
>         Return the discriminator loss.
>         We also call loss_D.backward() to calculate the gradients.
>         """
>         # Real
>         pred_real = netD(real)
>         loss_D_real = self.criterionGAN(pred_real, True)
>         # Fake
>         pred_fake = netD(fake.detach())
>         loss_D_fake = self.criterionGAN(pred_fake, False)
>         # wgan-gp
>         gradient_penalty, gradients = networks.cal_gradient_penalty(
>             netD, real, fake, self.device,lambda_gp=10.0
>         )
>         gradient_penalty.backward(retain_graph=True)
>         # Combined loss and calculate gradients
>         loss_D = (loss_D_real + loss_D_fake) * 0.5
>         loss_D.backward()
>         return loss_D
> ```
> 
> and to start training
> 
> ```
> python train.py --dataroot ./datasets/maps --name maps_cyclegan --model cycle_gan --gan_mode wgangp --norm instance 
> ```
> 
> Is this all? or should I take care of other things?
> Is this the same setup you tried ?

Hi @moh3th1 .
I wonder if your wgangp really works. If I use the original configuration, my networks just diverge. So, do you have some suggestions for hyperparameters of wgangp?

**Comment by jakubLangr** [comment]
I tried using WGAN-GP also on my own dataset. LSGAN does not produce good results so I tried using WGAN-GP, but not only are the results not better (I'd say worse), the losses seem to diverge to crazy values over time (1500+ at epoch 15). 

This work is fantastic, but I am somewhat surprised that there are no clear improvements to CycleGAN even after ~2.5 years? I'd love to try some techniques that are good candidates for improving upon CycleGAN. However, even bigger batchsize (cf. BigGAN) nor different image scales seem to help. Will continue investigating. But is anyone in this thread already looked at ways to improve? 

I'd be keen to explore e.g.:
https://github.com/akanimax/Variational_Discriminator_Bottleneck

**Comment by huyangdi** [comment]
> > 嗨，@ junyanz， 
> > 我想确保在使用相同设置和相同配置的情况下，尝试使用WGAN-GP丢失的CycleGAN。
> > 这是我受**cycle_gan_model.py**上的[＃439](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/issues/439)和[VON](https://github.com/junyanz/VON/blob/master/models/shape_gan_model.py#L80)启发的修改：****
> > ```
> > def backward_D_basic(self, netD, real, fake):
> >         """Calculate GAN loss for the discriminator
> > 
> >         Parameters:
> >             netD (network)      -- the discriminator D
> >             real (tensor array) -- real images
> >             fake (tensor array) -- images generated by a generator
> > 
> >         Return the discriminator loss.
> >         We also call loss_D.backward() to calculate the gradients.
> >         """
> >         # Real
> >         pred_real = netD(real)
> >         loss_D_real = self.criterionGAN(pred_real, True)
> >         # Fake
> >         pred_fake = netD(fake.detach())
> >         loss_D_fake = self.criterionGAN(pred_fake, False)
> >         # wgan-gp
> >         gradient_penalty, gradients = networks.cal_gradient_penalty(
> >             netD, real, fake, self.device,lambda_gp=10.0
> >         )
> >         gradient_penalty.backward(retain_graph=True)
> >         # Combined loss and calculate gradients
> >         loss_D = (loss_D_real + loss_D_fake) * 0.5
> >         loss_D.backward()
> >         return loss_D
> > ```
> > 
> > 
> > 并开始训练
> > ```
> > python train.py --dataroot ./datasets/maps --name maps_cyclegan --model cycle_gan --gan_mode wgangp --norm instance 
> > ```
> > 
> > 
> > 这是全部吗？还是我应该照顾其他事情？
> > 这是您尝试过的相同设置吗？
> 
> 嗨@ moh3th1。
> 我不知道您的帮派是否真的有效。如果使用原始配置，我的网络将出现分歧。那么，您对wgangp的超参数有什么建议吗？
> 
> 但是，似乎需要为wgan模型做些什么？

您的loss_d的写法是否存在问题，loss_D = (loss_D_real + loss_D_fake) * 0.5，0.5是您的超参数，但是是否应该为这个形式：loss_D = (loss_D_fake -loss_D_real) * 0.5

**Comment by huyangdi** [comment]
> > 嗨，@ junyanz， 
> > 我想确保在使用相同设置和相同配置的情况下，尝试使用WGAN-GP丢失的CycleGAN。
> > 这是我受**cycle_gan_model.py**上的[＃439](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix/issues/439)和[VON](https://github.com/junyanz/VON/blob/master/models/shape_gan_model.py#L80)启发的修改：****
> > ```
> > def backward_D_basic(self, netD, real, fake):
> >         """Calculate GAN loss for the discriminator
> > 
> >         Parameters:
> >             netD (network)      -- the discriminator D
> >             real (tensor array) -- real images
> >             fake (tensor array) -- images generated by a generator
> > 
> >         Return the discriminator loss.
> >         We also call loss_D.backward() to calculate the gradients.
> >         """
> >         # Real
> >         pred_real = netD(real)
> >         loss_D_real = self.criterionGAN(pred_real, True)
> >         # Fake
> >         pred_fake = netD(fake.detach())
> >         loss_D_fake = self.criterionGAN(pred_fake, False)
> >         # wgan-gp
> >         gradient_penalty, gradients = networks.cal_gradient_penalty(
> >             netD, real, fake, self.device,lambda_gp=10.0
> >         )
> >         gradient_penalty.backward(retain_graph=True)
> >         # Combined loss and calculate gradients
> >         loss_D = (loss_D_real + loss_D_fake) * 0.5
> >         loss_D.backward()
> >         return loss_D
> > ```
> > 
> > 
> > 并开始训练
> > ```
> > python train.py --dataroot ./datasets/maps --name maps_cyclegan --model cycle_gan --gan_mode wgangp --norm instance 
> > ```
> > 
> > 
> > 这是全部吗？还是我应该照顾其他事情？
> > 这是您尝试过的相同设置吗？
> 
> 嗨@ moh3th1。
> 我不知道您的帮派是否真的有效。如果使用原始配置，我的网络将出现分歧。那么，您对wgangp的超参数有什么建议吗？
> 
> 但是，似乎需要为wgan模型做些什么？

Is there a problem with the way you write loss_d? Loss_d = (loss_D_real + loss_D_fake) * 0.5， 0.5 is your hyperparameter, but should it be in this form: loss_d = (loss_d_fake-loss_d_real) * 0.5？

**Comment by AdrianoFonseca** [comment]
I believe in WGAN-GP you are not supposed to use any normalization in the Discriminator/Critic. Has anyone tried this with normalization in the Generator and not in the Discriminator?

**Comment by plutoyuxie** [comment]
> Is there a problem with the way you write loss_d? Loss_d = (loss_D_real + loss_D_fake) * 0.5， 0.5 is your hyperparameter, but should it be in this form: loss_d = (loss_d_fake-loss_d_real) * 0.5？

No need to change. The method self.criterionGAN has taken the sign into account.
